File: Solutions/Python/validate_binary_search_tree.py
"""
    [MEDIUM]
    98. Validate Binary Search Tree

    Concepts:
    - binary search tree
    - depth-first search

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 19.52 MB  [Beats 38.52%]
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        iot = []

        def dfs(curr):
            nonlocal iot

            if not curr:
                return

            if curr.left:
                dfs(curr.left)

            iot.append(curr.val)

            if curr.right:
                dfs(curr.right)

        dfs(root)
        
        for i in range(1, len(iot)):
            if iot[i-1] >= iot[i]:
                return False

        return True

File: Solutions/Cpp/validate_binary_search_tree.cpp
/*
    [MEDIUM]
    98. Validate Binary Search Tree

    Concepts:
    - binary search tree
    - depth-first search

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 22.20 MB  [Beats 9.62%]
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    std::vector<int> iot = {};

    bool dfs(TreeNode* curr) {
        if (!curr)
            return true;

        bool outcome = true;

        outcome &= dfs(curr->left);

        if (!iot.empty() && curr->val <= iot.back())
            return false;
        iot.push_back(curr->val);

        outcome &= dfs(curr->right);

        return outcome;
    }

    bool isValidBST(TreeNode* root) {
        return dfs(root);
    }
};